<a href="https://colab.research.google.com/github/atik-bd/MSc-BA/blob/main/Data%20preparation%20and%20Balancing%2002.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## APPLIED STATISTICS AND MACHINE LEARNING
## Exercise#1_ Atik Mahbub_ ID 20062447


# Data preparation (Breast Cancer dataset)

## Data Reading

In [ ]:
from pandas import read_csv,get_dummies,DataFrame
data1=read_csv('/content/Breast_Cancer_ 20062447.csv')

In [ ]:
data1.head()

Age   Race Marital Status T Stage  N Stage 6th Stage  \
0   68  White        Married       T1      N1       IIA   
1   50  White        Married       T2      N2      IIIA   
2   58  White       Divorced       T3      N3      IIIC   
3   58  White        Married       T1      N1       IIA   
4   47  White        Married       T2      N1       IIB   

               differentiate Grade   A Stage  Tumor Size Estrogen Status  \
0      Poorly differentiated     3  Regional           4        Positive   
1  Moderately differentiated     2  Regional          35        Positive   
2  Moderately differentiated     2  Regional          63        Positive   
3      Poorly differentiated     3  Regional          18        Positive   
4      Poorly differentiated     3  Regional          41        Positive   

  Progesterone Status  Regional Node Examined  Reginol Node Positive  \
0            Positive                      24                      1   
1            Positive                      14                      5   
2            Positive                      14                      7   
3            Positive                       2                      1   
4            Positive                       3                      1   

   Survival Months Status  
0               60  Alive  
1               62  Alive  
2               75  Alive  
3               84  Alive  
4               50  Alive

In [ ]:
data1.shape

(4024, 16)

In [ ]:
data1.describe()

Age   Tumor Size  Regional Node Examined  \
count  4024.000000  4024.000000             4024.000000   
mean     53.972167    30.473658               14.357107   
std       8.963134    21.119696                8.099675   
min      30.000000     1.000000                1.000000   
25%      47.000000    16.000000                9.000000   
50%      54.000000    25.000000               14.000000   
75%      61.000000    38.000000               19.000000   
max      69.000000   140.000000               61.000000   

       Reginol Node Positive  Survival Months  
count            4024.000000      4024.000000  
mean                4.158052        71.297962  
std                 5.109331        22.921430  
min                 1.000000         1.000000  
25%                 1.000000        56.000000  
50%                 2.000000        73.000000  
75%                 5.000000        90.000000  
max                46.000000       107.000000

In [ ]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Age                     4024 non-null   int64 
 1   Race                    4024 non-null   object
 2   Marital Status          4024 non-null   object
 3   T Stage                 4024 non-null   object
 4   N Stage                 4024 non-null   object
 5   6th Stage               4024 non-null   object
 6   differentiate           4024 non-null   object
 7   Grade                   4024 non-null   object
 8   A Stage                 4024 non-null   object
 9   Tumor Size              4024 non-null   int64 
 10  Estrogen Status         4024 non-null   object
 11  Progesterone Status     4024 non-null   object
 12  Regional Node Examined  4024 non-null   int64 
 13  Reginol Node Positive   4024 non-null   int64 
 14  Survival Months         4024 non-null   int64 
 15  Stat

## Data Cleaning
 In our dataset all 16 columns have equal number of datapoints, 4024 entries. So, data cleaning is not required here.

## Data Encoding

In [ ]:
data1['A Stage']
data1['Estrogen Status']
data1['Progesterone Status']
data1['Status']

0       Alive
1       Alive
2       Alive
3       Alive
4       Alive
        ...  
4019    Alive
4020    Alive
4021    Alive
4022    Alive
4023    Alive
Name: Status, Length: 4024, dtype: object

In [ ]:
data1['T Stage ']=data1['T Stage '].map({'T1':1,'T2':2,'T3':3,'T4':4})
data1['N Stage']=data1['N Stage'].map({'N1':1,'N2':2,'N3':3})
data1['6th Stage']=data1['6th Stage'].map({'IIA':1,'IIIA':2,'IIB':3,'IIIB':4,'IIIC':5})
data1['differentiate']=data1['differentiate'].map({'Undifferentiated':1,'Poorly differentiated':2,'Moderately differentiated':3,'Well differentiated':4})
data1['Grade']=data1['Grade'].map({'1':1,'2':2,'3':3,' anaplastic; Grade IV':4})

In [ ]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Age                     4024 non-null   int64 
 1   Race                    4024 non-null   object
 2   Marital Status          4024 non-null   object
 3   T Stage                 4024 non-null   int64 
 4   N Stage                 4024 non-null   int64 
 5   6th Stage               4024 non-null   int64 
 6   differentiate           4024 non-null   int64 
 7   Grade                   4024 non-null   int64 
 8   A Stage                 4024 non-null   object
 9   Tumor Size              4024 non-null   int64 
 10  Estrogen Status         4024 non-null   object
 11  Progesterone Status     4024 non-null   object
 12  Regional Node Examined  4024 non-null   int64 
 13  Reginol Node Positive   4024 non-null   int64 
 14  Survival Months         4024 non-null   int64 
 15  Stat

In [ ]:
data1['A Stage']=data1['A Stage'].map({'Regional':1,'Distant':0})
data1['Estrogen Status']=data1['Estrogen Status'].map({'Positive':1,'Negative':0})
data1['Progesterone Status']=data1['Progesterone Status'].map({'Positive':1,'Negative':0})
data1['Status']=data1['Status'].map({'Alive':1,'Dead':0})

In [ ]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Age                     4024 non-null   int64 
 1   Race                    4024 non-null   object
 2   Marital Status          4024 non-null   object
 3   T Stage                 4024 non-null   int64 
 4   N Stage                 4024 non-null   int64 
 5   6th Stage               4024 non-null   int64 
 6   differentiate           4024 non-null   int64 
 7   Grade                   4024 non-null   int64 
 8   A Stage                 4024 non-null   int64 
 9   Tumor Size              4024 non-null   int64 
 10  Estrogen Status         4024 non-null   int64 
 11  Progesterone Status     4024 non-null   int64 
 12  Regional Node Examined  4024 non-null   int64 
 13  Reginol Node Positive   4024 non-null   int64 
 14  Survival Months         4024 non-null   int64 
 15  Stat

In [ ]:
data2=get_dummies(data1,['Race','Marital Status'],dtype=int)

In [ ]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   Age                       4024 non-null   int64
 1   T Stage                   4024 non-null   int64
 2   N Stage                   4024 non-null   int64
 3   6th Stage                 4024 non-null   int64
 4   differentiate             4024 non-null   int64
 5   Grade                     4024 non-null   int64
 6   A Stage                   4024 non-null   int64
 7   Tumor Size                4024 non-null   int64
 8   Estrogen Status           4024 non-null   int64
 9   Progesterone Status       4024 non-null   int64
 10  Regional Node Examined    4024 non-null   int64
 11  Reginol Node Positive     4024 non-null   int64
 12  Survival Months           4024 non-null   int64
 13  Status                    4024 non-null   int64
 14  Race_Black                4024 non-null 

In [ ]:
data2.to_csv('Breast Cancer Data_Atik.csv')

## Defining X and Y

In [ ]:
y=data2['Status']
x=data2.drop('Status',axis=1)
print (x.shape)
print (y.shape)

(4024, 21)
(4024,)


## Scaling X

In [ ]:
from sklearn.preprocessing import StandardScaler
x_scaled=StandardScaler().fit_transform(x)
DataFrame(x_scaled)

0         1         2         3         4         5         6   \
0     1.565253 -1.025287 -0.632209 -1.057202 -1.331031  1.331031  0.152963   
1    -0.443222  0.281158  0.809974 -0.269245  0.235987 -0.235987  0.152963   
2     0.449434  1.587604  2.252157  2.094626  0.235987 -0.235987  0.152963   
3     0.449434 -1.025287 -0.632209 -1.057202 -1.331031  1.331031  0.152963   
4    -0.777968  0.281158 -0.632209  0.518712 -1.331031  1.331031  0.152963   
...        ...       ...       ...       ...       ...       ...       ...   
4019  0.895761 -1.025287 -0.632209 -1.057202  0.235987 -0.235987  0.152963   
4020  0.226270  0.281158  0.809974 -0.269245  0.235987 -0.235987  0.152963   
4021  1.565253  0.281158 -0.632209  0.518712  0.235987 -0.235987  0.152963   
4022  0.449434  0.281158 -0.632209  0.518712  0.235987 -0.235987  0.152963   
4023 -0.889550  0.281158 -0.632209  0.518712  0.235987 -0.235987  0.152963   

            7         8         9   ...        11        12        13  \
0    -1.253661  0.267652  0.458107  ... -0.618172 -0.492961 -0.279201   
1     0.214345  0.267652  0.458107  ...  0.164807 -0.405695 -0.279201   
2     1.540287  0.267652  0.458107  ...  0.556296  0.161530 -0.279201   
3    -0.590691  0.267652  0.458107  ... -0.618172  0.554224 -0.279201   
4     0.498475  0.267652  0.458107  ... -0.618172 -0.929288 -0.279201   
...        ...       ...       ...  ...       ...       ...       ...   
4019 -1.016886  0.267652  0.458107  ... -0.618172 -0.972921 -0.279201   
4020  0.735251  0.267652  0.458107  ...  0.752041 -0.100266 -0.279201   
4021 -0.401271  0.267652 -2.182898  ... -0.226682 -0.100266 -0.279201   
4022  0.640541  0.267652  0.458107  ... -0.618172  0.030632  3.581645   
4023 -0.022430  0.267652  0.458107  ... -0.422427  1.252348 -0.279201   

            14        15        16        17        18        19        20  
0    -0.293927  0.423109 -0.370629  0.722850 -0.106346 -0.424741 -0.249041  
1    -0.293927  0.423109 -0.370629  0.722850 -0.106346 -0.424741 -0.249041  
2    -0.293927  0.423109  2.698117 -1.383413 -0.106346 -0.424741 -0.249041  
3    -0.293927  0.423109 -0.370629  0.722850 -0.106346 -0.424741 -0.249041  
4    -0.293927  0.423109 -0.370629  0.722850 -0.106346 -0.424741 -0.249041  
...        ...       ...       ...       ...       ...       ...       ...  
4019  3.402205 -2.363456 -0.370629  0.722850 -0.106346 -0.424741 -0.249041  
4020 -0.293927  0.423109  2.698117 -1.383413 -0.106346 -0.424741 -0.249041  
4021 -0.293927  0.423109 -0.370629  0.722850 -0.106346 -0.424741 -0.249041  
4022 -0.293927 -2.363456  2.698117 -1.383413 -0.106346 -0.424741 -0.249041  
4023 -0.293927  0.423109 -0.370629  0.722850 -0.106346 -0.424741 -0.249041  

[4024 rows x 21 columns]

## Data Splitting

In [ ]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x_scaled,y,test_size=0.25, random_state=67)
print (x_train.shape)
print (x_test.shape)
print (y_train.shape)
print (y_test.shape)

(3018, 21)
(1006, 21)
(3018,)
(1006,)


## Data Balancing (for Training set)

In [ ]:
y_train.value_counts()

Status
1    2546
0     472
Name: count, dtype: int64

In [ ]:
from imblearn.over_sampling import SMOTE
x_train,y_train=SMOTE(random_state=103).fit_resample(x_train,y_train)

In [ ]:
y_train.value_counts()

Status
0    2546
1    2546
Name: count, dtype: int64